adbe/adb_enhanced.py

```diff
@@ -207,24 +207,48 @@ def handle_layout(value):
 
 
 # Source: https://stackoverflow.com/questions/10506591/turning-airplane-mode-on-via-adb
-# This is incomplete
 def handle_airplane(turn_on):
+    return_code_wifi, output_wifi, _ = execute_adb_shell_settings_command2("get global wifi_on")
+    return_code_data, output_data, _ = execute_adb_shell_settings_command2("get global mobile_data")
+
+    if return_code_wifi != 0:
+        print_error('Failed to get wifi state')
+        return _USER_PRINT_VALUE_UNKNOWN
+
+    if return_code_data != 0:
+        print_error('Failed to get mobile-data state')
+        return _USER_PRINT_VALUE_UNKNOWN
+
     if turn_on:
-        cmd = 'put global airplane_mode_on 1'
-    else:
-        cmd = 'put global airplane_mode_on 0'
-
-    # At some version, this became a protected intent, so, it might require root to succeed.
-    broadcast_change_cmd = 'am broadcast -a android.intent.action.AIRPLANE_MODE'
-    # This is a protected intent which would require root to run
-    # https://developer.android.com/reference/android/content/Intent.html#ACTION_AIRPLANE_MODE_CHANGED
-    broadcast_change_cmd = 'su root %s' % broadcast_change_cmd
-    execute_adb_shell_settings_command2(cmd)
-    return_code, _, _ = execute_adb_shell_command2(broadcast_change_cmd)
-    if return_code != 0:
-        print_error_and_exit('Failed to change airplane mode')
+        return_code_wifi, _, _ = execute_adb_shell_settings_command2("put global adbe_wifi " + output_wifi)
+        return_code_data, _, _ = execute_adb_shell_settings_command2("put global adbe_data " + output_data)
+        return_code_airplane, _, _ = execute_adb_shell_settings_command2("put global airplane_mode_on 1")
+
+        if return_code_wifi != 0 or return_code_data != 0 or return_code_airplane != 0:
+            print_error('Failed to put "Global" settings states. Proceeding anyway ...')
+
+        handle_mobile_data(False)
```

Comment:
Why not send a protected intent as well?
It will just fail in most cases but can be good when it succeeds.